### ISSUE TITLE
PHP Notice:  Malformed file attributes

### ISSUE THREAD

**Issue by PHLAK** [phpseclib/phpseclib#1708]
I'm getting the following error/warning/notice after this package was updated to `2.0.34`.

```php
>>> Storage::disk('sftp')->put('./test/file.test', 'Test file; please ignore')
<warning>PHP Notice:  Malformed file attributes in {LARAVEL_APP_PATH}/vendor/phpseclib/phpseclib/phpseclib/Net/SFTP.php on line 3162</warning>
<warning>PHP Notice:  Malformed file attributes in {LARAVEL_APP_PATH}/vendor/phpseclib/phpseclib/phpseclib/Net/SFTP.php on line 3162</warning>
=> true
```

Specifically, I'm using Laravel's `Storage` facade with the `league/flysystem-sftp` package that has a dependency of `"phpseclib/phpseclib": "~2.0"`.

I've temporarily worked around the issue by adding the following to my composer `require` section.

```
"phpseclib/phpseclib": "^2.0 !=2.0.34",
```

Also, this only seemed to be an issue with PHP 8, I wasn't getting the error from a PHP 7.4 container.

**Comment by terrafrost** [comment]
Can you post the SFTP logs? You can get them by doing `define('NET_SFTP_LOGGING', 2);` at the top and I guess `Storage::disk('sftp')->getSFTPLog()` after you get the error.

If `Storage::disk('sftp')->getSFTPLog()` doesn't work try `define('NET_SFTP_LOGGING', 3);`.

Thanks!

**Comment by terrafrost** [comment]
phpseclib 2.0.34 introduced SFTPv4/5/6 support but it _should_ default to SFTPv3 unless you _explicitly_ tell it to use SFTPv3.

This is relevant because the error appears to be in the `_parseTime` method, which _should_ only be called when SFTPv4/5/6 is being used.

Now, I suppose it could be called if NET_SFTP_ATTR_CREATETIME was erroneously being used as an attribute for a v3 server  but still....  the logs would be very helpful to figure that out.

Thanks!

**Comment by PHLAK** [comment]
I will attempt to grab these for you next week.

**Comment by danijelk** [comment]
> Can you post the SFTP logs? You can get them by doing `define('NET_SFTP_LOGGING', 2);` at the top and I guess `Storage::disk('sftp')->getSFTPLog()` after you get the error.
> 
> If `Storage::disk('sftp')->getSFTPLog()` doesn't work try `define('NET_SFTP_LOGGING', 3);`.
> 
> Thanks!

Here are our logs, same error, thanks.

```


-> NET_SFTP_STAT (0.0001s)
00000000  00:00:00:25:2f:42:75:74:69:6b:73:6c:69:73:74:65  ...%/Butiksliste
00000010  2f:44:61:6e:69:63:61:43:65:6e:74:72:65:42:75:74  /DanicaCentreBut
00000020  69:6b:6b:65:72:2e:78:6c:73                       ikker.xls



<- NET_SFTP_ATTRS (0.0323s)
00000000  00:00:00:3d:00:00:00:00:00:05:50:00:00:00:81:e8  ...=......P.....
00000010  5f:a3:8e:47:61:87:b1:b7                          _..Ga...


<warning>PHP Notice:  Malformed file attributes in /app/vendor/phpseclib/phpseclib/phpseclib/Net/SFTP.php on line 3162</warning>
<warning>PHP Notice:  Malformed file attributes in /app/vendor/phpseclib/phpseclib/phpseclib/Net/SFTP.php on line 3162</warning>

```

**Comment by terrafrost** [comment]
Those aren't the full SFTP logs. As is it's unclear what version of SFTP is being used. The first few lines of the SFTP logs should look like this:

```
-> NET_SFTP_INIT (0.0002s)
00000000  00:00:00:03                                      ....

<- NET_SFTP_VERSION (0.1428s)
00000000  00:00:00:03:00:00:00:18:70:6f:73:69:78:2d:72:65  ........posix-re
00000010  6e:61:6d:65:40:6f:70:65:6e:73:73:68:2e:63:6f:6d  name@openssh.com
00000020  00:00:00:01:31:00:00:00:13:73:74:61:74:76:66:73  ....1....statvfs
00000030  40:6f:70:65:6e:73:73:68:2e:63:6f:6d:00:00:00:01  @openssh.com....
00000040  32:00:00:00:14:66:73:74:61:74:76:66:73:40:6f:70  2....fstatvfs@op
00000050  65:6e:73:73:68:2e:63:6f:6d:00:00:00:01:32:00:00  enssh.com....2..
00000060  00:14:68:61:72:64:6c:69:6e:6b:40:6f:70:65:6e:73  ..hardlink@opens
00000070  73:68:2e:63:6f:6d:00:00:00:01:31:00:00:00:11:66  sh.com....1....f
00000080  73:79:6e:63:40:6f:70:65:6e:73:73:68:2e:63:6f:6d  sync@openssh.com
00000090  00:00:00:01:31                                   ....1

-> NET_SFTP_REALPATH (0.0002s)
00000000  00:00:00:01:2e                                   .....
```
So with SFTP you have the main version and then you have other versions that can be negotiated. I'm almost wondering if your main version _isn't_ v3 but without the above snippet I can't make that determination.

Also, it'd be helpful to know what SSH software you're connecting to. You can get that by doing `echo $sftp->getServerIdentification()`.

That said...

```
<- NET_SFTP_ATTRS (0.0323s)
00000000  00:00:00:3d:00:00:00:00:00:05:50:00:00:00:81:e8  ...=......P.....
00000010  5f:a3:8e:47:61:87:b1:b7                          _..Ga...
```
The first 4 bytes (00:00:00:3d) represent the flags. The flags you have are as follows:

- SSH_FILEXFER_ATTR_SIZE
- SSH_FILEXFER_ATTR_PERMISSIONS
- SSH_FILEXFER_ATTR_ACCESSTIME
- SSH_FILEXFER_ATTR_CREATETIME
- SSH_FILEXFER_ATTR_MODIFYTIME

The latter two weren't introduced until SFTPv4. So the two possibilities are:

1. Your server is using SFTPv4 syntax even tho it's using SFTPv3 (at which point it'd be a bug with your server)
2. Your server is using SFTPv4 and phpseclib is assuming it's SFTPv3 (at which point it'd be a phpseclib bug)

My guess is that maybe it's the latter. That the server's main version isn't SFTPv3 but rather v4. I'll do some tests later today to see if that could cause this issue. BUT the full logs would tell me _conclusively_ if that's what was going on.

Also, can you tell me what SSH server your running by doing `echo $sftp->getServerIdentification()`? I'm guessing you're _not_ connecting to OpenSSH. If I'm not able to set the main SFTP version in the Windows SFTP server that I have it might be beneficial to try another one. Like maybe the one you're using (assuming you're not using OpenSSH; OpenSSH, in-so-far as I know, doesn't support any SFTP version other than v3)

**Comment by danijelk** [comment]
Ah didn't think they were relevant, below is a pre `"phpseclib/phpseclib": "^2.0 !=2.0.34",` and after (when it works). Only seems to be a diff in the NET_SFTP_ATTRS flags as you mentioned. I can't use the getServerIdentification method on a flysystem disk() interface but I got this from telnet output, hope that can be used SSH-2.0-WS_FTP-SSH_7.7

```

-> NET_SFTP_INIT (0.0001s)
00000000  00:00:00:03                                      ....



<- NET_SFTP_VERSION (0.0141s)
00000000  00:00:00:03                                      ....



-> NET_SFTP_REALPATH (0s)
00000000  00:00:00:01:2e                                   .....



<- NET_SFTP_NAME (0.0597s)
00000000  00:00:00:01:00:00:00:01:2f:00:00:00:35:64:72:77  ......../...5drw
00000010  78:72:2d:78:2d:2d:2d:20:32:20:66:74:70:61:64:6d  xr-x--- 2 ftpadm
00000020  69:6e:20:66:74:70:61:64:6d:69:6e:20:20:20:20:20  in ftpadmin
00000030  20:20:20:30:20:4a:61:6e:20:30:31:20:31:39:37:30     0 Jan 01 1970
00000040  20:2f:00:00:00:04:00:00:01:e8                     /........



-> NET_SFTP_STAT (0.0001s)
00000000  00:00:00:25:2f:42:75:74:69:6b:73:6c:69:73:74:65  ...%/Butiksliste
00000010  2f:44:61:6e:69:63:61:43:65:6e:74:72:65:42:75:74  /DanicaCentreBut
00000020  69:6b:6b:65:72:2e:78:6c:73                       ikker.xls



<- NET_SFTP_ATTRS (0.0323s)
00000000  00:00:00:3d:00:00:00:00:00:05:50:00:00:00:81:e8  ...=......P.....
00000010  5f:a3:8e:47:61:87:b1:b7                          _..Ga...


<warning>PHP Notice:  Malformed file attributes in /app/vendor/phpseclib/phpseclib/phpseclib/Net/SFTP.php on line 3162</warning>
<warning>PHP Notice:  Malformed file attributes in /app/vendor/phpseclib/phpseclib/phpseclib/Net/SFTP.php on line 3162</warning>

-> NET_SFTP_OPEN (0.0001s)
00000000  00:00:00:25:2f:42:75:74:69:6b:73:6c:69:73:74:65  ...%/Butiksliste
00000010  2f:44:61:6e:69:63:61:43:65:6e:74:72:65:42:75:74  /DanicaCentreBut
00000020  69:6b:6b:65:72:2e:78:6c:73:00:00:00:01:00:00:00  ikker.xls.......
00000030  00                                               .
```

```


-> NET_SFTP_INIT (0s)
00000000  00:00:00:03                                      ....



<- NET_SFTP_VERSION (0.0127s)
00000000  00:00:00:03                                      ....



-> NET_SFTP_REALPATH (0s)
00000000  00:00:00:01:2e                                   .....



<- NET_SFTP_NAME (0.0313s)
00000000  00:00:00:01:00:00:00:01:2f:00:00:00:35:64:72:77  ......../...5drw
00000010  78:72:2d:78:2d:2d:2d:20:32:20:66:74:70:61:64:6d  xr-x--- 2 ftpadm
00000020  69:6e:20:66:74:70:61:64:6d:69:6e:20:20:20:20:20  in ftpadmin
00000030  20:20:20:30:20:4a:61:6e:20:30:31:20:31:39:37:30     0 Jan 01 1970
00000040  20:2f:00:00:00:04:00:00:01:e8                     /........



-> NET_SFTP_STAT (0s)
00000000  00:00:00:25:2f:42:75:74:69:6b:73:6c:69:73:74:65  ...%/Butiksliste
00000010  2f:44:61:6e:69:63:61:43:65:6e:74:72:65:42:75:74  /DanicaCentreBut
00000020  69:6b:6b:65:72:2e:78:6c:73                       ikker.xls



<- NET_SFTP_ATTRS (0.0312s)
00000000  00:00:00:3d:00:00:00:00:00:05:4e:00:00:00:81:e8  ...=......N.....
00000010  5f:a3:8e:47:61:89:1f:58                          _..Ga..X



-> NET_SFTP_OPEN (0.0001s)
00000000  00:00:00:25:2f:42:75:74:69:6b:73:6c:69:73:74:65  ...%/Butiksliste
00000010  2f:44:61:6e:69:63:61:43:65:6e:74:72:65:42:75:74  /DanicaCentreBut
00000020  69:6b:6b:65:72:2e:78:6c:73:00:00:00:01:00:00:00  ikker.xls.......
00000030  00                                               .

```

**Comment by PHLAK** [comment]
Here's a dump of the logs I was able to generate.

```
```

**Comment by terrafrost** [comment]
Well I'm not able to install WS_FTP server. When I try I get a dumb `WS_FTP Server is not supported on Workstations or Domain Controllers. Aborting setup` error (I'm trying to install it on Windows 10 Pro).

According to https://docs.ipswitch.com/WS_FTP_Server76/Help/Server/index.htm?system_requirements.htm?toc.htm I guess it may not be supported on Windows 10.

In lieu of that...  try this branch:

https://github.com/terrafrost/phpseclib/tree/issue1708

**Comment by danijelk** [comment]
> Well I'm not able to install WS_FTP server. When I try I get a dumb `WS_FTP Server is not supported on Workstations or Domain Controllers. Aborting setup` error (I'm trying to install it on Windows 10 Pro).
> 
> According to https://docs.ipswitch.com/WS_FTP_Server76/Help/Server/index.htm?system_requirements.htm?toc.htm I guess it may not be supported on Windows 10.
> 
> In lieu of that... try this branch:
> 
> https://github.com/terrafrost/phpseclib/tree/issue1708

Works here, no warnings are thrown anymore. Thanks for your support.

**Comment by terrafrost** [comment]
I've merged that into the 1.0 / 2.0 / 3.0 / master branches:

https://github.com/phpseclib/phpseclib/commit/40035dc59c33fdf32d716b0804495104c54b5105

Also, fwiw, I'd class this as a bug with WS_FTP Server. They should *not* be setting bits for attributes that aren't supported by SFTPv3. But whatever - we got a workaround that works!
